### ISSUE TITLE
Explain UID mapping for folder sharing

### ISSUE THREAD

**Issue by techtonik** [lxc/lxd#733]
This is related to #714. But in order to resolve it, I miss some details. First of all, LXD defines a ranges of UID that container manager is allowed to to use on host. Then it selects the UID to run the guest. Then I need to mount the folder so that guest and host users both can access to it.

Should I create some group on host that includes my current user id on host and guest UID and set privileges on shared folder to be group writable?

Are there other ways?

**Comment by stgraber** [comment]
Group membership on the host is meaningless in this case as the uid_t entry for the container's task will not be in those groups.

Your options are really either to chown to a uid that's mapped into the container, chgrp to a gid that's mapped into the container, make the directory readable/writable to everyone or use POSIX ACLs.

Also, the container doesn't have a single UID and GID assigned to it, but 65536 of each. Currently all containers share the same range (the first 65536 uids and gids of the lxd user's allocation) but this may change in the future.

**Comment by techtonik** [comment]
So, to ensure that all my files are accessible inside of container, I need 3 things:

1. `bind mount` directory inside of container
2. figure you UID of guest user on host machine
3. set POSIX ACL so that guest user UID has full access to all the files

The most important question - will the guest user be able to break the ACL if it uses chown/chmod instead of getfacl/setfacl? But how-to in points 1 and 2 are also unclear.

**Comment by stgraber** [comment]
Since this was first opened we did two pieces of work that should help most users:
 - Made it easy to map host uid/gid directly into a container (https://stgraber.org/2017/06/15/custom-user-mappings-in-lxd-containers/)
 - Introduced storage pools and storage volumes where storage volumes can be attached to one or more containers and has LXD automatically remap uid/gid in that case.

We're also looking at kernel work for mount based uid/gid mapping (such as shiftfs) but at this time no such capabilities have been added to the Linux kernel. As soon as that happens, we'll have LXD make use of it for all filesystem operations, making it possible to share data between host and container and between containers without ever having to care about ownership issues.
